File: src/core/src/main/kotlin/com/cshep4/premierpredictor/component/fixtures/FixtureFormatter.kt
package com.cshep4.premierpredictor.component.fixtures

import com.cshep4.premierpredictor.data.Match
import com.cshep4.premierpredictor.data.api.live.match.MatchFacts
import org.springframework.stereotype.Component


@Component
class FixtureFormatter {
    private val remappingFunction = { v1: List<Match>, v2: List<Match> ->
        val list = v1.toMutableList()
        list.addAll(v2)
        list
    }

    fun format(fixturesApiResult: List<MatchFacts>): List<Match> {
        return fixturesApiResult.map { it.toMatch() }
    }

    fun groupIntoTeams(matches: List<Match>): Map<String, List<Match>> {
        val homeTeamMatches = matches.groupBy { it.hTeam }
        val awayTeamMatches = matches.groupBy { it.aTeam }

        val allMatches = homeTeamMatches.toMutableMap()

        awayTeamMatches.forEach { k, v -> allMatches.merge(k, v, remappingFunction) }

        return allMatches.map { Pair(it.key, it.value.sortedBy { v -> v.dateTime }) }.toMap()
    }
}

File: src/fixtureservice/lib/component/fixture-formatter.ts
import {Fixture} from "../model/fixture";
import {asSequence} from "sequency";

export class FixtureFormatter {
    public groupIntoTeams(fixtures: Fixture[]): Promise<Map<string, Fixture[]>> {
        return new Promise<Map<string, Fixture[]>>((resolve, reject) => {
            const homeTeamMatches: Map<string, Fixture[]> = asSequence(fixtures)
                .groupBy(f => f.hTeam);
            const awayTeamMatches: Map<string, Fixture[]> = asSequence(fixtures)
                .groupBy(f => f.aTeam);

            const allMatches = homeTeamMatches;

            awayTeamMatches.forEach((v: Fixture[], k: string) => {
                allMatches.set(k, this.mergeArrays(k, allMatches, awayTeamMatches));

                const sortedFixtures: Fixture[] = asSequence<Fixture>(allMatches.get(k))
                    .sortedBy((f: Fixture) => {
                        return f.dateTime
                    })
                    .toArray();

                allMatches.set(k, sortedFixtures);
            });

            resolve(allMatches);
        });
    }

    private mergeArrays(k: string, allMatches: Map<string, Fixture[]>, awayTeamMatches: Map<string, Fixture[]>): Fixture[] {
        if (!allMatches.has(k)) {
            return awayTeamMatches.get(k);
        }

        return asSequence<Fixture>(allMatches.get(k))
            .merge(awayTeamMatches.get(k), (f: Fixture) => {
                return f.id;
            })
            .toArray();
    }
}